posts/2022-04-15-libertydevops-gitjenkinshelmiks.adoc

```diff
@@ -0,0 +1,180 @@
+                              ibmcloud ks cluster config --cluster $CLUSTER-ID
+                              kubectl config current-context
+                              kubectl create -f deploy/deployment.yaml #( simple k8s deployment command )
+                              kubectl create -f deploy/service.yaml #( simple k8s deployment command )
+                              kubectl create -f deploy/route.yaml #( simple k8s deployment command )
+                           '''
+
+Make sure that your Kubernetes configuration files are in the same source code repository (under a folder called "deploy"), also the image name should be updated manually or programmatically 
+
+image::/img/blog/deploymentyaml.png[Image reference in deployment yaml,width=40%,align="center"]
+
+
+When Jenkins has checked out the Java application code for the code build, all the Kubernetes configuration files are also downloaded to the Jenkins workspace so that Jenkins can run the IBM Cloud and Kubernetes commands to connect to the Kubernetes cluster and deploy the application.
+
+All the other needed application deployment commands are available in this Kubernetes command page that is straightforward 
+link:https://kubernetes.io/docs/reference/kubectl/cheatsheet/[Kubernetes sample commands].
+
+== QA testing, Jakarta EE & MicroProfile applications with Jenkins
+
+Apart from running JUnit test cases along with the code build phase, Jenkins can trigger functional and integration QA test cases automatically after deploying the Jakarta EE & MicroProfile application.
+
+Configure the test cases on jenkins job and test it manually. Create an "Authentication Token" in "Trigger builds remotely" section under "Build Triggers". Trigger this test case from docker "entrypoint" file by using remote rest API call that uses this authentication token as the identifier
+
+For example: curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<authentication-token>
+Note: Auth token can be generated from postman
+
+== Kubernetes monitoring tools
+
+Kubernetes provides commands to check the application or cluster logs and memory/cpu usage through the commands like 
+
+    $ kubectl logs ..
+    $ cat /sys/fs/cgroup/cpu/cpuacct.usage (after connecting to k8s pod)
+    $ cat /sys/fs/cgroup/memory/memory.usage_in_bytes (after connecting to k8s pod)
+
+For persistence of logs and usage stats, several applications are available in the market, which can be integrated with Kubernetes, details about some of those apps are given below
+
+These tools are deployed in Kubernetes cluster itself where the application is running and exposed by using route and access the gathered details from UI.
+
+    -> https://grafana.com/oss/loki/
+    -> https://medium.com/nerd-for-tech/logging-at-scale-in-kubernetes-using-grafana-loki-3bb2eb0c0872
+    -> https://prometheus.io
+    -> https://k21academy.com/docker-kubernetes/prometheus-grafana-monitoring/
+
```

Comment:
Rephrase to:

"There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://grafana.com/oss/loki/[Grafana] and link:https://prometheus.io[Prometheus]. You can read more in the following articles:

* link:https://medium.com/nerd-for-tech/logging-at-scale-in-kubernetes-using-grafana-loki-3bb2eb0c0872[ARTICLE TITLE HERE]
* link:https://k21academy.com/docker-kubernetes/prometheus-grafana-monitoring/[ARTICLE TITLE HERE]
"

Reply:
Done